### ISSUE TITLE
SSP offset feature test

### ISSUE THREAD

**Issue by Blebowski** [Blebowski/CTU-CAN-FD#208]
Implement feature test for secondary sampling point user defined offset.
All options of SW offset, measured offset, SW + measured offset must be executed.

**Assignees:** furavikt

*Migrated from GitLab: https://gitlab.fel.cvut.cz/canbus/ctucanfd_ip_core/-/issues/205*

**Comment by Blebowski** [comment]
This needs Data and Secondary sampling points in Status Bus:
https://gitlab.fel.cvut.cz/canbus/ctucanfd_ip_core/issues/260

**Comment by Blebowski** [comment]
Hi,
I added 3 sample signals to Status Bus in:
https://gitlab.fel.cvut.cz/canbus/ctucanfd_ip_core/issues/260

These are:
1. Sample trigger for Nominal Bit Time delayed by 1 clock cycle.
2. Sample trigger for Data Bit Time delayed by 1 clock cycle.
3. Sample trigger for Secondary Sample delayed by 1 clock cycle.

Little Design description:
If Core is transmitter, it is using Nominal Sampling point up to Bit-rate switch in EDL(FDF) bit.
After the switch, it will start using Data-Sample. Secondary sample point is invisible to
CAN Core (it is used only in Bus sampling module. CAN Core is still processing receive data
with Data Sample. To make sure that correct data are processed (since RX data might not be
yet there due to roundtrip), data into bit-destuffing are driven from post bit stuffing data.

This-way, Protocol Control receives what it transmitted in the Same bit. Note that this does
not mind since after bit-rate switch Core is the only transmitter.

Now to the error handling. If the Core which is transmitting in Data-Bit rate is automatically
receiving what it transmitted, then it has no way of knowing what has happened on the bus
(what errors are there). But what types of Errors does transmitter in Data-Phase needs
to detect? It will detect Bit-errors in Bus Sampling via secondary sample point. It should detect,
Stuff errors if some other unit starts transmitting Error Frame. But if this happends, the unit
will anyway detect Bit-Error sooner or later (due to stuff bit).
To sum it up: Transmitter in Data phase needs to only detect bit errors and it will even recognize
error frame by bit error detection mechanism.

Now to the test:
I would propose to wait for a Bit-rate switch (sp_control signal is also available in Status Bus),
and then measure number of clock cycles it takes for first occurence of Secondary Sample Point.

**Comment by Blebowski** [comment]
***In GitLab by @furavikt on Feb 26, 2019, 20:56 GMT+1:***

created branch [`205-ssp-offset-feature-test`](https://gitlab.fel.cvut.cz/canbus/ctucanfd_ip_core/compare/master...205-ssp-offset-feature-test) to address this issue

**Comment by Blebowski** [comment]
closed via merge request !306

**Comment by Blebowski** [comment]
mentioned in commit 72ab7315b99fea3cc5e870c8de9b72c8c003a2e0

**Comment by Blebowski** [comment]
reopened

**Comment by Blebowski** [comment]
closed via merge request !311

**Comment by Blebowski** [comment]
mentioned in commit 089604543e797722354187c9fd13a3c93db05ba2

**Comment by Blebowski** [comment]
reopened

**Comment by Blebowski** [comment]
reopened

**Comment by Blebowski** [comment]
reopened

**Comment by Blebowski** [comment]
mentioned in commit 262777dd5cf002889e4a4f204faa8687dbfb1a8a

**Comment by Blebowski** [comment]
reopened

**Comment by Blebowski** [comment]
mentioned in commit c07dd74fb31d9d34a7b5def6e3009aa4c66a3af9
